### ISSUE TITLE
VIPs are not readded after recreating a tun/tap interface

### ISSUE THREAD

**Issue by vnxme** [opnsense/core#3972]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[x] I have read [the contributing guide lines](https://github.com/opnsense/core/blob/master/CONTRIBUTING.md).
[x] I have searched the existing issues and I'm convinced that mine is new.

**Describe the bug**
VIPs are not readded after recreating a tun/tap interface. I experienced this problem when using Tinc but it seems to be related to any tun/tap interface so I generalised it.

**To Reproduce**
1. Add a network and at least one host in Tinc configuration
2. Assign a newly created Tinc interface
3. Add a virtual IP (alias) to the interface
4. Restart Tinc daemon from the dashboard or make any changes in Tinc configuration
5. Check that the virtual IP is missing

**Expected behavior**
Virtual IPs should be readded each time a tun/tap interface is created.

**Environment**
OPNsense 20.1.2

**Comment by AdSchellevis** [comment]
It's the plugins responsibility to handle its own interfaces, openvpn for example triggers the configd action `interface newip {interface}`  in ovpn-linkup, which is called from 
https://github.com/opnsense/core/blob/ff290c95ec27b7afb306ab63295ccd390da1ea18/src/etc/inc/plugins.inc.d/openvpn.inc#L607

Same applies to dhcp type interfaces and ppp links. If an interface signals a devd event, these will be catched as well, using https://github.com/opnsense/core/blob/eb95e8f53e56e25ce8f0898ee6936d8201bd9a29/src/etc/devd/ifnet.conf

(calling rc.linkup which also boils down to the same event)
https://github.com/opnsense/core/blob/eb95e8f53e56e25ce8f0898ee6936d8201bd9a29/src/etc/rc.linkup#L60

**Comment by vnxme** [comment]
@AdSchellevis, just checked it manually: if I call `/usr/local/opnsense/service/configd_ctl.py interface newip $tinc_interface` after Step 5 above, a VIP is not readded. Probably I have misunderstood your point. Otherwise adding such a line in tincd,py could easily resolve the problem.

**Comment by AdSchellevis** [comment]
@vnxme that should work indeed, maybe we're missing something here. If I have some time available, I'll try to setup something similar on my end.

**Comment by vnxme** [comment]
@AdSchellevis, thanks in advance, your contribution is very much appreciated.

BTW, when Tinc is restarted, it is not only a VIP that disappears but also a network route. E.g. Tinc interface address is fc00::1/120 and a virtual IP is 10.0.0.1/24. When I restart Tinc daemon or change configuration, `netstat -rn4` does not have 10.0.0.1/24 at all and `netstat -rn6` has only fc00::1/128 (assigned to lo0) while the network route /120 is missing.

To fix network route /120 I have to do the following:
- Remove a VIP (otherwise a red warning is displayed on the interface page)
- Reassign the interface (go to Interfaces -> Tinc_Interface and click Save)
- Restart Tinc daemon

**Comment by AdSchellevis** [comment]
@vnxme hmm, on my end it works like a charm, below my output:

````
root@OPNsense:~/core # ifconfig tinc0
tinc0: flags=8043<UP,BROADCAST,RUNNING,MULTICAST> metric 0 mtu 1500
	options=80000<LINKSTATE>
	inet 192.168.100.1 netmask 0xffff0000 broadcast 192.168.255.255 
	inet6 fe80::20c:29ff:feaf:8196%tinc0 prefixlen 64 scopeid 0x22 
	nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
	groups: tun tinc 
	Opened by PID 48813
root@OPNsense:~/core # configctl interface newip tinc0
OK
root@OPNsense:~/core # ifconfig tinc0
tinc0: flags=8043<UP,BROADCAST,RUNNING,MULTICAST> metric 0 mtu 1500
	options=80000<LINKSTATE>
	inet 192.168.100.1 netmask 0xffff0000 broadcast 192.168.255.255 
	inet 192.168.123.130 netmask 0xffffffff broadcast 192.168.123.130 
	inet6 fe80::20c:29ff:feaf:8196%tinc0 prefixlen 64 scopeid 0x22 
	nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
	groups: tun tinc 
	Opened by PID 48813
````

`192.168.123.130 `  is the address configured as vip:
![image](https://user-images.githubusercontent.com/9783985/76332456-b2396380-62f0-11ea-8302-7bdc265f87fc.png)

**Comment by vnxme** [comment]
@AdSchellevis, thank you for investing your time in it. I suppose, you a) used tun, not tap; b) used primary IPv4 (192.168.100.1 in your example); c) used same address family VIP (192.168.123.130 in your example). Although it does not sound reasonable, all these factors make a difference.

I spent another couple of hours experimenting with it and the problem I described seems to be a little more complex than just a configd_ctl.py call... Tomorrow I will try to document every detail to show you.

My first preliminary guesses are as follows:
1. There are some issues with TAP interfaces, e.g. no fe80::/10 link locals are created for them.
2. There is an issue with configd_ctl.py script when called on interface with primary IPv6 address: `opnsense: /usr/local/etc/rc.newwanip: Failed to detect IP for TINC0[opt1]` message in log.
3. There are some issues with IPv6 network routes: they disappear from time to time leaving only /128 address assigned to lo0 in netstat -rn6 output which breaks routing.

**Comment by AdSchellevis** [comment]
@vnxme the `newipv6` is probably the one you're looking for if you would like to signal IPv6, in my case that would be `configctl interface newipv6 tinc0` ppp interfaces seem to do the same for ipv6 only.

Maybe I can reproduce your issue after you supplied the additional information, I expect you only need some additions in tinc to make this functional.

**Comment by vnxme** [comment]
The table below illustrates all the issues I identified in connection with the readding VIPs problem:
1. Unlike TUN interfaces, TAP interfaces do not have IPv6 link-local net+address pairs (fe80) though both have auto_linklocal option enabled - see red cells F37:G58;
2. Kernel/OPNsense loses (or fails to add) IPv6 networks while adding IPv6 link-local and primary net+address pairs - see red cells G14:G22, K26:K34, K50:K58 - it consequently breaks routing;
3. Kernel/OPNsense keeps (or fails to delete) IPv6 networks while deleting IPv6 alias net+address pairs - see red cells O21, O33 and O57;
4. Kernel/OPNsense keeps (or fails to delete) IPv4 alias net+address pair and IPv6 alias address for TAP interface while deleting corresponding net+address pairs - see red cells L42:M42 and L45:N45;
5. Configd_ctl.py fails to readd IPv6 alias network for TAP with primary IPv4 at step 10 - see red cell O46.

Unless you want me to open separate issue reports for the aforementioned facts, I suggest we should continue discussing them here.

![Capture-1](https://user-images.githubusercontent.com/46669194/76401223-9dada780-6392-11ea-8399-fde27e09a3ee.PNG)
![Capture-2](https://user-images.githubusercontent.com/46669194/76401333-c33ab100-6392-11ea-98d5-2cbc8fb9d04c.PNG)

**Comment by AdSchellevis** [comment]
@vnxme Most of your issues seem to be related to tap interfaces combined with ipv6, tap is hardly used on our end. 

Just one question, if you do the following, it results in a tap interface without link local on your end (if I understand your chart correctly):

- create new tinc network in switch mode
- add a couple of hosts
- apply settings

On my end it seems to create an interface with link local in this case. I'm afraid it will be quite labor intensive to debug your problem and at the moment I'm quite convinced that the issue isn't in core, but either a missing piece in Tinc or an upstream problem if it's very specific to tap interfaces.

By my knowledge we don't have similar reports about OpenVPN, which uses tun devices as well.

The tun plugin is something I wrote a long time ago for a customer, who eventually switched to a different setup, so unfortunately we're not really working on it ourselves at the moment (other than reviewing PR's and fixing minor bugs).

**Comment by vnxme** [comment]
@AdSchellevis Yes, issues 1, 4 and 5 seem to be related to tap, but issues 2 and 3 are related to primary and virtual IPv6 addresses. The most crucial is the second one as it breaks routing (with no IPv6 network on interface packets are sent by OPNsense using default route which is wrong).

Yes, you are right, a tap interface ~~is~~ was created without link-local address. In my case the same ~~is~~ was true not only for Tinc (mode switch), but also for OpenVPN (mode tap) and Zerotier. **Update:** Actually I traced back this particular issue with tap interfaces to `net.link.tap.up_on_open=1`. I added it as it had been recommended by [Zerotier pkg-message](https://www.freshports.org/net/zerotier/). I have removed it from System/Settings/Tunables and restarted, so tap interfaces are created with link-local addresses now.

I will rebuild the chart for switch (tap) cases shortly. Besides, I am going to propose an alteration to tincd.py as a part of [PR 1733](https://github.com/opnsense/plugins/pull/1733) so that tinc-up scripts properly trigger configd_ctl.py. Nevertheless, issues 2 and 3 are still valid.

**Comment by vnxme** [comment]
@AdSchellevis Now, as we have an updated picture, it looks more or less symmetric in terms of tun/tap. Link-local addresses now work as expected. 

As for issue 2, the pattern appears to provide for Kernel/OPNsense improper handling of the last added IPv6 addresses on interfaces
- It handles link-local addresses properly (does not delete /64 net) when an interface has a primary IPv6 address (the primary is the last added one as fe80 is always added first);
- It does not handle link-local addresses properly when it is the only IPv6 address, i.e. when an interface has a primary IPv4 address.

Issues 3 and 4 do not break anything for me but something is definitely wrong with how kernel/OPNsense handles VIPs.

**Comment by AdSchellevis** [comment]
on FreeBSD / IPv6, primary is a concept that doesn't really exist, some context about this can be found here https://github.com/opnsense/core/issues/3310.

Your statement about something being completely wrong in handing VIPs, I don't share, haven't seen proof of that to be honest and other interfaces implementing the correct events don't seem to be suffering from it.

If you can write steps how to reproduce the specific issue, it might be better to open a new ticket for that and close this one. 

I can't spend more time on this subject, unless there is a clear direction on where it should lead.

**Comment by vnxme** [comment]
> on FreeBSD / IPv6, primary is a concept that doesn't really exist, some context about this can be found here #3310.

No, it does not exist. However, I believe it's a good way to make a distinction between an address assigned by the corresponding daemon (I indicated on the chart that primary addresses are assigned by Tinc in my case) and an alias address assigned by some firewall scripts. Haven't realy found anything related to the issue I described in #3310.

> Your statement about something being completely wrong in handing VIPs, I don't share, haven't seen proof of that to be honest and other interfaces implementing the correct events don't seem to be suffering from it.

Sorry, I did not intend to hurt you or anyone else by my statement. I provided you with the results of my tests showing that e.g. when in switch mode at step 9 (tincd restart) the system's netstat -rn4 and -rn6 output have VIPs while they should not be there (false positives).

> If you can write steps how to reproduce the specific issue, it might be better to open a new ticket for that and close this one.

It's a good idea, I will do that.

> I can't spend more time on this subject, unless there is a clear direction on where it should lead.

Thank you again for your time and effort!

**Comment by AdSchellevis** [comment]
>  Sorry, I did not intend to hurt you or anyone else by my statement. I provided you with the results of my tests showing that e.g. when in switch mode at step 9 (tincd restart) the system's netstat -rn4 and -rn6 output have VIPs while they should not be there (false positives).

No offence taken, just saying there's not enough to point to a general core issue yet.

**Comment by vnxme** [comment]
@AdSchellevis You are right it is not a core issue. I managed to resolve the missing IPv6 network route issue (number 2 and I believe it also includes issues 3 and 4 above) by destroying tun/tap interface each time Tinc stops/restarts. Suggested changes are in [PR 1733](https://github.com/opnsense/plugins/pull/1733).
